### DISCUSSION TITLE
In python formulae, what is the difference between `depends_on` and a `resource` for the same PyPI package?

### DISCUSSION THREAD

**Discussion by unitof** [Homebrew/discussions#5211] [Tap maintenance and Homebrew development] [answered]
What (if any) is the difference in a formula between, say

```
depends_on "cython"
```

and

```
resource "cython" do
  url "https://files.pythonhosted.org/packages/0e/17/c5b026cea7a634ee3b8950a7be16aaa49deeb3b9824ba5e81c13ac26f3c4/Cython-3.0.9.tar.gz"
  sha256 "a2d354f059d1f055d34cfaa62c5b68bc78ac2ceab6407148d47fb508cf3ba4f3"
end
```
for a formula which depends on cython from PyPI?

Is it always redundant to have both? If so, which is preferred?

(Context: I am working on a [manim formula](https://github.com/Homebrew/homebrew-core/pull/166434) and trying to clean up my resources/dependencies.)

## Update/Example

In testing, I've discovered using `depends_on "pygments"` in my formula does not work. The formula builds but fails `brew test manim` with:

```
==> Testing manim
==> /opt/homebrew/Cellar/manim/HEAD-9f3d486/bin/manim -ql /private/tmp/manim-test-20240319-38046-b2toy4/testscene.py CreateCircle
Last 15 lines from /Users/jacob/Library/Logs/Homebrew/manim/test.01.manim:

Traceback (most recent call last):
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/bin/manim", line 5, in <module>
    from manim.__main__ import main
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/libexec/lib/python3.12/site-packages/manim/__init__.py", line 13, in <module>
    from ._config import *
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/libexec/lib/python3.12/site-packages/manim/_config/__init__.py", line 10, in <module>
    from .logger_utils import make_logger
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/libexec/lib/python3.12/site-packages/manim/_config/logger_utils.py", line 23, in <module>
    from rich.logging import RichHandler
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/libexec/lib/python3.12/site-packages/rich/logging.py", line 15, in <module>
    from .traceback import Traceback
  File "/opt/homebrew/Cellar/manim/HEAD-9f3d486/libexec/lib/python3.12/site-packages/rich/traceback.py", line 23, in <module>
    from pygments.lexers import guess_lexer_for_filename
ModuleNotFoundError: No module named 'pygments'
Error: manim: failed
```

However, removing `depends_on "pygments"` and switching to a resource block

```
resource "pygments" do
  url "https://files.pythonhosted.org/packages/55/59/8bccf4157baf25e4aa5a0bb7fa3ba8600907de105ebc22b0c78cfbf6f565/pygments-2.17.2.tar.gz"
  sha256 "da46cec9fd2de5be3a8a784f434e4c4ab670b4ff54d605c4c2717e9d49c4c367"
end
```

builds and passes `brew test manim`. Both the homebrew & PyPI versions of `pygments` are presently identical: `2.17.2`.

**Comment by SMillerDev** [comment, marked as answer]
> Is it always redundant to have both? If so, which is preferred?

Yes, since they represent the same dependency. The `depends_on` is preferred since it will automatically allow updates without bumping the formula.
Homebrew as a project is reducing the duplication between python resources and formulae though, by removing formulae that do not provide much benefit over resources: https://github.com/Homebrew/homebrew-core/issues/157500

**Reply by unitof** [reply]
Thanks! So would the issue I ran into (`depends_on "pygments"` working while `resource "pygments"` fails in the `manim` formula) represent a bug with either pygments or its formula?

Or, are there valid circumstances where the resource is required over the Homebrew formula?
